tests/integration/v0/test_integration.py

```diff
@@ -2,19 +2,243 @@
+        )
+        await ops_test.model.deploy(
+            TestIntegration.requirer_charm,
+            application_name=f"{IP_ROUTER_REQUIRER_APP_NAME}-a",
+            series="jammy",
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[f"{IP_ROUTER_REQUIRER_APP_NAME}-a", IP_ROUTER_PROVIDER_APP_NAME],
+            status="blocked",
+            timeout=1000,
+        )
+
+    @pytest.mark.abort_on_fail
+    async def test_given_charms_deployed_when_relate_then_status_is_active(self, ops_test):
+        await ops_test.model.add_relation(
+            relation1=f"{IP_ROUTER_REQUIRER_APP_NAME}-a",
+            relation2=IP_ROUTER_PROVIDER_APP_NAME,
+        )
+
+        await ops_test.model.wait_for_idle(
+            apps=[f"{IP_ROUTER_REQUIRER_APP_NAME}-a", IP_ROUTER_PROVIDER_APP_NAME],
+            status="active",
+            timeout=1000,
+        )
+
+        # Check that the router has the correct app name in the routing table
+        provider_unit = ops_test.model.units[f"{IP_ROUTER_PROVIDER_APP_NAME}/0"]
+
+        action = await provider_unit.run_action(action_name="get-routing-table")
+        action_output = await ops_test.model.get_action_output(
+            action_uuid=action.entity_id, wait=60
+        )
+        assert json.loads(action_output["msg"]) == {f"{IP_ROUTER_REQUIRER_APP_NAME}-a": []}
+
+    @pytest.mark.abort_on_fail
+    async def test_given_network_request_provider_implements_and_requirer_sees(self, ops_test):
+        # This emits an event called new_network_request
+        provider_unit = ops_test.model.units[f"{IP_ROUTER_PROVIDER_APP_NAME}/0"]
+        requirer_unit = ops_test.model.units[f"{IP_ROUTER_REQUIRER_APP_NAME}-a/0"]
+
+        # Run a "request-network" action on the requirer charm
+        requested_network = [{"network": "192.168.250.0/24", "gateway": "192.168.250.1"}]
+        action = await requirer_unit.run_action(
+            action_name="request-network", network=json.dumps(requested_network)
+        )
+        action_output = await ops_test.model.get_action_output(
+            action_uuid=action.entity_id, wait=60
+        )
+        assert action_output["msg"] == "ok"
+
+        # Wait for the model to finish executing `relation-changed`
+        await ops_test.model.wait_for_idle(
+            apps=[f"{IP_ROUTER_REQUIRER_APP_NAME}-a", IP_ROUTER_PROVIDER_APP_NAME],
+            status="active",
+            timeout=1000,
+        )
+
+        # Run a "get-routing-table" action on the provider charm
+        provider_unit = ops_test.model.units[f"{IP_ROUTER_PROVIDER_APP_NAME}/0"]
+        action = await provider_unit.run_action(action_name="get-flattened-routing-table")
+        action_output = await ops_test.model.get_action_output(
+            action_uuid=action.entity_id, wait=60
+        )
+        assert json.loads(action_output["msg"]) == requested_network
+
+    @pytest.mark.abort_on_fail
+    async def test_full_end2end_test(self, ops_test):
```

Comment:
Recommend validating only one functionality at a time
Need to rename the test by using given, when and then words

Reply:
Changed the name to fit the given/when/then paradigm, since I would consider this the single, main functionality.